sdk/webpubsub/azure-webpubsub-client/azure/webpubsub/client/_client.py

```diff
@@ -0,0 +1,858 @@
+        :type data_type: Any.
+        :keyword bool ack: If False, the message won't contains ackId and no AckMessage
+         will be returned from the service. Default is True.
+        :keyword bool no_echo: Whether the message needs to echo to sender. Default is False.
+        """
+
+        def send_to_group_attempt():
+            ack = kwargs.pop("ack", True)
+            no_echo = kwargs.pop("no_echo", False)
+            if ack:
+                self._send_message_with_ack_id(
+                    message_provider=lambda id: SendToGroupMessage(
+                        group=group_name, data_type=data_type, data=content, ack_id=id, no_echo=no_echo
+                    ),
+                    **kwargs,
+                )
+            else:
+                self._send_message(
+                    message=SendToGroupMessage(group=group_name, data_type=data_type, data=content, no_echo=no_echo),
+                    **kwargs,
+                )
+
+        self._retry(send_to_group_attempt)
+
+    def _retry(self, func: Callable[[], None]):
+        retry_attempt = 0
+        while True:
+            try:
+                func()
+                return
+            except Exception as e:  # pylint: disable=broad-except
+                retry_attempt = retry_attempt + 1
+                delay_seconds = self._message_retry_policy.next_retry_delay(retry_attempt)
+                if delay_seconds is None:
+                    raise e
+                _LOGGER.debug("will retry %sth times after %s seconds", retry_attempt, delay_seconds)
+                time.sleep(delay_seconds)
+
+    def _call_back(self, callback_type: Union[CallbackType, str], *args):
+        for func in self._handler[callback_type]:
+            # _call_back works in listener thread which must not be blocked so we have to execute the func
+            # in new thread to avoid dead lock
+            threading.Thread(target=func, args=args, daemon=True).start()
+
+    def _start_from_restarting(self):
+        if self._state != WebPubSubClientState.DISCONNECTED:
+            _LOGGER.warning("Client can be only restarted when it's Disconnected")
+            return
+
+        try:
+            self._start_core()
+        except Exception as e:
+            self._state = WebPubSubClientState.DISCONNECTED
+            raise e
+
+    def _handle_auto_reconnect(self):
+        _LOGGER.debug("start auto reconnect")
+        success = False
+        attempt = 0
+        while not self._is_stopping:
+            try:
+                self._start_from_restarting()
+                success = True
+                break
+            except Exception as e:  # pylint: disable=broad-except
+                _LOGGER.warning("An attempt to reconnect connection failed %s", e)
+                attempt = attempt + 1
+                delay_seconds = self._reconnect_retry_policy.next_retry_delay(attempt)
+                if not delay_seconds:
+                    break
```

Comment:
Should this be `break` or `continue`?

Reply:
It shall be "break". `delay_seconds` is None which means client shall not retry anymore.